### ISSUE TITLE
in yaml, a ~ key should allow to assing a value to the parent key

### ISSUE THREAD

**Issue by rmanibus** [FasterXML/jackson-dataformats-text#506]
In Yaml spec, if one property is the prefix of another, a value can be attributed to the parent one using ~ as a null key to represent any YAML property that is a prefix of another one.

The following test is failing:

```
    public void testTildeIsDeserializedAsNullKey() throws Exception
    {
        final String YAML =
                "quarkus:\n" +
                        "  log:\n" +
                        "    sentry:\n" +
                        "      ~: true\n" +
                        "      dsn: 'some string'";

        JsonNode node = MAPPER.readValue(YAML, JsonNode.class);
        assertTrue(node.get("quarkus").get("log").get("sentry").booleanValue());
    }
```

instead, the following statement is true:

`assertTrue(quarkus.get("quarkus").get("log").get("sentry").get("~").booleanValue())`

quarkus.log.sentry."~" is true instead of quarkus.log.sentry

Relevant conversation:

https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/yaml.20property.20conflict

Relevant spec:

https://quarkus.io/guides/config-yaml#configuration-property-conflicts

https://yaml.org/spec/1.2.2/

**Comment by rmanibus** [comment]
To clarify, this is causing issues with the Spotless formatter, it is basically transforming any ~ key in the "~" string

**Comment by cowtowncoder** [comment]
Quick question: how would you represent such document without `~` marker? Wouldn't there be problem with `dsn` key conflicting -- that is, there being both:

* quarkus.log.sentry = true
* quarkus.log.sentry.dsn = "some string"

which is impossible structurally?

Put another way; how would equivalent, non-tilde-using document look like?

**Comment by rmanibus** [comment]
I think you cannot, that's the whole point of the ~ key, it allows you to assign a value to a key that also has children

**Comment by cowtowncoder** [comment]
So could you not access `quarkus.log.sentry.dsn` in such case? Or is that simply dropped in such a case?

If there is no equivalent document, I don't think Jackson data model can represent this.
And if not, this is something that cannot really be supported -- it might actually be best to just throw an exception if so.

But I guess I am also not clear on point of using such marker.
I think I am missing something obvious.

**Comment by cowtowncoder** [comment]
Ohhh. Maybe it's effectively "empty" key? So something like JSON would have:

```json
{ "quarkus": {
      "log": {
         "sentry": {
            "": true,
            "dsn": "some string"
         }
      }
   }
}
```

... although that would not be hidden by Jackson either; would need to be access with empty String as key.

**Comment by rmanibus** [comment]
The point of using this marker is to allow the use of yaml in configuration systems that have not initially been thought of with yaml in mind

**Comment by cowtowncoder** [comment]
Yeah the problem is that it does really map to `JsonNode` model (nor can I see an easy way to model it). In your case, either "null" key would need to replace value of `sentry` (so "sentry.dsn" would be lost), or we'd need a surrogate for "null" key.
For latter, empty String seems close enough -- but I am not sure TBH whether it'd be any more convenient to access than "~" key.

Put another way: what is being specifically asked is not doable. But I am open to ideas of alternative improvements.

**Comment by cowtowncoder** [comment]
Correct: Jackson's streaming/token model is based on JSON and there's no `null` key, only `null` values. And `JsonNode`, similarly only `String` keys. YAML and other formats need to be expressed in these models.

Theoretically I suppose it would be possible to expose `null` key but I don't think it'd work through all processing.

And even if it was done, would require access by `null` as key like:

    node.get("quarkus").get("log").get("sentry").get(null).booleanValue()

**Comment by maxandersen** [comment]
Yeah; but that (using null) for lookup would be expected.  But yeah I can see how that could be problematic in places.

**Comment by cowtowncoder** [comment]
I think I am still inclined to prefer empty String as surrogate (and possible make this `YAMLParser.Feature` whether to replace "~" key or not -- since there may well be existing code that relies on current handling).

I was first thinking of allowing specific "replacement key", that'd also allow `null`, but I think that there are too many places downstream that wouldn't stomach `null` that it's probably not worth it (more work to have non-boolean switches, but more importantly, trying to make databind pieces accept `null` for `JsonToken.FIELD_NAME` and so on).

**Comment by asomov** [comment]
@maxandersen the `~`  character represents null in YAML 1.1, in YAML 1.2 they made YAML a superset of JSON (which does not support this char as null). [The types have becode optional ](https://yaml.org/type/null.html)(they depend on the schema).
It is possible to configure SnakeYAML Engine to respect `~` as you wish, I am not sure Jackson should be able to do it.

**Comment by maxandersen** [comment]
I grok the concern - its just that with this it is not possible in yaml to assign values to the "parent". Thats why in Quarkus (and other) yaml configs the notion of ~ (not "~") - that works fine; becuse as you state in yaml 1.2 the schemas/usecase can choose to interpret it as null or not. We do that in Quarkus using snakeyaml - all is fine.

Problem comes when tools like spotless uses jackson to parse yaml thinking that jackson as fully yaml compliant but the formatting now "mutates" the content.

Hence it would IMO be nice that jackson at least had some way to retain the info.

**Comment by asomov** [comment]
@maxandersen I am afraid, there is some confusion here:
- As far as I can see Quarkus  is giving `null` a very special meaning
- I think Jackson 3.0 is fully YAML 1.2 compliant (because it is using [SnakeYAML Engine](https://bitbucket.org/snakeyaml/snakeyaml-engine))
- Quarkus does not use YAML 1.2 - I wonder why your examples refer to YAML 1.2
- Quarkus is using [SnakeYAML](https://bitbucket.org/snakeyaml/snakeyaml) (which explicitly supports YAML 1.1, and it does not define schemas)
- I can help to migrate Quarkus to SnakeYAML Engine (and YAML 1.2)

It looks like the issue was reported for Jackson 2.x (which is also using SnakeYAML and it is YAML 1.1 compliant)

**Comment by maxandersen** [comment]
@asomov i'm not (afaics) referring to YAML 1.2 - but even with that afaics Yaml 1.2 spec still specifies "null | Null | NULL | ~" as equivalent ... and https://yaml.org/type/null.html seem to confirm it to me.

hence i'm not sure what you mean that Quarkus would need to define a schema to get this behavior?

**Comment by asomov** [comment]
1. Quarkus does not need to do anything if it is happy with what it has (YAML 1.1)
2. Quarkus would need to define a schema only if it wants to use YAML 1.2 and apply SnakeYAML Engine
3. YAML 1.2 defines `~` as `null` as a part of [the Core schema](https://yaml.org/spec/1.2.2/#103-core-schema) (as far as I remember Jackson 3.0 uses the JSON schema at the moment)
4. Jackson serves a very general purpose, and probably[ the JSON schema ](https://yaml.org/spec/1.2.2/#102-json-schema)better matches its use cases for the community.

**Comment by maxandersen** [comment]
...the issue is that spotless formatter uses jackson where ~ are not retained but converted into "~" ...not sure how that can be fixed without jackson being able to handle null ..?

**Comment by asomov** [comment]
@maxandersen if the expectration to support `~` is for Jackson 3 (and YAML 1.2) then I do not support that expectation.
I hope Qarkus may find a better solution for `null` representation. (or to support the internal functional requirement)
Using the Core YAML Schema would lead to even more questions for a general purpose parser.
Jackson may consider to implement a way to explicitly configure the Core Schema, but I think it will not help you - your users will not configure Jackson.

**Comment by maxandersen** [comment]
@asomov I feel we are misunderstanding each other here :)

Quarkus is working just fine - this yaml is valid yaml in any linter/parser/spec:

```
log:
  sentry:
    ~: true
    dsn: "some string"
```

The only reason I'm speaking about it here is that *spotless* uses jackson to parse the yaml and format/validate with it and here jackson mutates the output.

I don't follow what Quarkus can do differently here besides maybe also go "around" the jackson and snakeyaml-engine behavior and be ok with "" empty string to mean null...?

**Comment by asomov** [comment]
@maxandersen yes, it is valid, but it may result in different object. For YAML 1.1 (which is inside Quarcus), `~` is it parsed to `null`. For YAML 1.2 it may be parsed differently depending on the context. For absolute majority of users `~` is just a char (that is why YAML 1.2 became a superset of JSON).
Please note that Jackson does not **mutate** anything! It properly parses the input to create **expected** output.

**Comment by asomov** [comment]
@maxandersen Quarcus may drop the usage of specific feature of YAML 1.1 and the issue disappears.

**Comment by cowtowncoder** [comment]
I guess from my end, while I think I understand the issue, I am not sure how to address this at Jackson level.

The original request is something I do not think can be supported, unfortunately (seamless access via `JsonNode`), due to `JsonNode` being incapable of expressing such structures (like JSON format itself).
But something else perhaps could be supported -- if this is, f.ex, about just avoiding quoting `~` String (forcing "raw" output) to support round-tripping: not sure this is possible using SnakeYAML (Jackson 2.x) or snakeyaml-engine (3.x) but might be.

So I am open to figuring out something that can be done.

**Comment by maxandersen** [comment]
@asomov again - Quarkus is just fine here - its the reformatting happening when spotless use jackson of rewriting the content that is problematic IMO. Which is what @cowtowncoder is also hinting at.

I don't know enough about snakeyaml engine to know yet if that is possible.

It is unfortunate yaml 1.2 don't allow it as it makes yaml unnecessary verbose compared to a properties file :)

**Comment by asomov** [comment]
@maxandersen I am completely confused when you say "_rewriting the content_". It means that something is already written, and Jackson replaces it with something else. I see how Jackson can parse, but I do not get how it may "_replace_"

**Comment by cowtowncoder** [comment]
@asomov I think he means round-tripping (read-write), in which something reads YAML with Jackson into, say, `JsonNode`, possibly modifies some parts, than writes out as YAML. If so, "~" as key will be represented differently (as quoted String) as a result.
At least that is my reading.

If so, option I am thinking of (a new `YAMLWriteFeature`) is something like `YAMLWriteFeature.WRITE_TILDE_KEY_UNQUOTED` (to write '~' key unquoted).

Bigger change, if I had time to really go through the whole process, would be thinking of a way to support actual "null keys" for `JsonNode`. But that'd be much farther out at this point (if ever).

**Comment by asomov** [comment]
@cowtowncoder it cannot be the case because SnakeYAML does not wrire `null` as `~ `(unless explicitly instructed, but it is not the case, I studied the source code of Quarkus and I did not see any explicit representer.)
`YAMLWriteFeature.WRITE_TILDE_KEY_UNQUOTED` cannot be used without switching the schema from JSON to Core.
@maxandersen can you please clarify "_rewriting the content_"?

**Comment by maxandersen** [comment]
@asomov @cowtowncoder is spot on - the issue is that spotless reads the yaml with jackson and writes it out again - in this process the null symbol is redefined to be a String so when it gets written back its with double quotes.

**Comment by asomov** [comment]
@maxandersen proposed solutions:

1. implement a fix in spotless (to apply the Core schema using SnakeYAML Engine). But I wonder how should spotless know which schema to apply
2. stop using` ~ `as the `null` representation. Stick to the standard null representation

**Comment by maxandersen** [comment]
What is the standard as null?

**Comment by cowtowncoder** [comment]
> [@cowtowncoder](https://github.com/cowtowncoder) it cannot be the case because SnakeYAML does not wire 
> `null` as `~ `(unless explicitly instructed, but it is not the case, I studied the source code of Quarkus and
> I did not see any explicit representer.) `YAMLWriteFeature.WRITE_TILDE_KEY_UNQUOTED` cannot be used 
> without switching the schema from JSON to Core. 
>
> [@maxandersen](https://github.com/maxandersen) can you please clarify "_rewriting the content_"?

Just to make sure: this is talking about `~` as object key, not as value: as far as I know, use as regular value works. But the Object key case is trickier: Jackson token model (and `JsonNode`) has no way to model that usage.

And because of this I do think (... but may still be wrong :) ) that `~` must become a `String` key when read by Jackson; and further will not be written as plain `~` but as quoted String `"~"`.
Whereas reading and writing `~` value it is exposed and written as proper null value in YAML.

**Comment by mjcarman** [comment]
> In Yaml spec, if one property is the prefix of another, a value can be attributed to the parent one using ~ as a null key to represent any YAML property that is a prefix of another one.

YAML allows null values as keys, but it **does not** apply their corresponding value to the the parent node — which wouldn't necessarily be a mapping, and wouldn't even exist at the document root. Even if the parent was a mapping, you'd end up with two values for one key.

> I think you cannot, that's the whole point of the ~ key, it allows you to assign a value to a key that also has children

> The point of using this marker is to allow the use of yaml in configuration systems that have not initially been thought of with yaml in mind

You're mistaking [Quarkus' application of YAML](https://quarkus.io/guides/config-yaml#configuration-property-conflicts) for a feature of YAML. 

In your example:
```
quarkus:
  log:
    sentry:
      ~: true
      dsn: 'some string'
```

Quarkus is (internally) converting this to:

```
quarkus.log.sentry: true
quarkus.log.sentry.dsn: 'some string'
```
as if it had been contained in a Java `.properties` file. That is; it's flattening the hierarchical YAML into a single-level map of key/value pairs.

**Comment by asomov** [comment]
@maxandersen sorry, I did not see your question.
[Indeed, it is confusing](https://yaml.org/type/null.html). Canonical is `~`

**Comment by maxandersen** [comment]
> > In Yaml spec, if one property is the prefix of another, a value can be attributed to the parent one using ~ as a null key to represent any YAML property that is a prefix of another one.
> 
> YAML allows null values as keys, but it **does not** apply their corresponding value to the the parent node — which wouldn't necessarily be a mapping, and wouldn't even exist at the document root. Even if the parent was a mapping, you'd end up with two values for one key.
> 
> > I think you cannot, that's the whole point of the ~ key, it allows you to assign a value to a key that also has children
> 
> > The point of using this marker is to allow the use of yaml in configuration systems that have not initially been thought of with yaml in mind
> 
> You're mistaking [Quarkus' application of YAML](https://quarkus.io/guides/config-yaml#configuration-property-conflicts) for a feature of YAML. 

As I wrote elsewhere. Quarkus is fine. Please ignore anything regarding quarkus here. Quarkus uses snakeyaml so fastjackson issue is not relevant for quarkus. 

It's the spotlessformatter that is not able to format valid yaml because Jackson has a lossy interpretation for ~.

**Comment by cowtowncoder** [comment]
@maxandersen Ok thank you yes, Quarkus handling not issue. But -- I hope I finally got it -- round-trip handling (read, modify, write) by Jackson YAML module is, since it produces different output. 
And mutation seen is quoting of `~` as `"~"` which changes semantics of the key to be just regular String key, not "null" key?
Is that correct understanding?

Now, how `~` is represented internally may be problematic too, but not the immediate concern.

How to solve the primary issue is the question then. One straight-forward way -- assuming snakeyaml-engine supports it -- would be to have `YAMLWriteFeature` for writing `String` `"~"` key as unquoted, null key.
That should likely work for most use cases as it seems unlikely that "~-as-a-String" is commonly needed.
And is easier than trying to expose "null-ness" of key itself.

But before working on that, quick question @rmanibus / @maxandersen : how exactly is Jackson used -- is YAML read by `ObjectMapper.readTree()`, then possibly modified, written out with `ObjectMapper.writeValue[AsString]`? Or using some other methods?

**Comment by mjcarman** [comment]
> [Indeed, it is confusing](https://yaml.org/type/null.html). Canonical is `~`

`~` is canonical in YAML 1.1. `null` is canonical in YAML 1.2, with `~` being recognized as an alternate representation.

I think there's a [schema](https://yaml.org/spec/1.2.2/#chapter-10-recommended-schemas) problem here. Under the Core Schema, `~` is resolved as `!!null` (and thus means `null`). Under the JSON Schema it's resolved as `!!str` (and thus means `"~"`); only `null` is resolved as a null value.

Jackson is currently hard-coded to use JSON schema. It will be possible to override that in 3.2.0, but in that case jackson would end up seeing a `null` key while reading input.

**Comment by asomov** [comment]
@mjcarman Jackson 2 is YAML 1.1 and it does not support schemas (because it is using SnakeYAML). Same for Quarkus.
Jackson 3 is YAML 1.2 (because it is using SnakeYAML Engine). When Quarkus creates a YAML 1.1 document, which is parsed by YAML 1.2 parser, it may get unexpected results.

**Comment by cowtowncoder** [comment]
Getting `null` for Object key will probably not work well as things are, for Jackson 3.x.

Since `JsonToken` based streams are not allowed to pass `null` Property Names, would likely need to map to String `"~"`...

**Comment by cowtowncoder** [comment]
@mjcarman and just to make sure -- as far as I know, handling of `~` (etc) as value is not the issue here; use as Object/Map key is.
